File: sesion5-lab/aspnet-jwt-attacks/jwt-demo/Controllers/JwtController.cs
using JwtDemo.Services;
using Microsoft.AspNetCore.Mvc;

namespace JwtDemo.Controllers;

[ApiController]
[Route("api/jwt")]
public class JwtController : ControllerBase
{
    private readonly VulnerableJwtService _vulnerable;
    private readonly SecureJwtService _secure;

    public JwtController(VulnerableJwtService vulnerable, SecureJwtService secure)
    {
        _vulnerable = vulnerable;
        _secure = secure;
    }

    [HttpPost("vulnerable/issue")]
    public IActionResult IssueVulnerable([FromBody] Dictionary<string, string> body)
    {
        var subject = body.GetValueOrDefault("subject", "user1");
        var role = body.GetValueOrDefault("role", "USER");
        return Ok(new
        {
            modo = "VULNERABLE",
            token = _vulnerable.IssueToken(subject, role),
            hint = "Prueba alg=none: eyJhbGciOiJubmUifQ.eyJzdWIiOiJhZG1pbiIsInJvbGUiOiJBRE1JTiJ9."
        });
    }

    [HttpGet("vulnerable/verify")]
    public IActionResult VerifyVulnerable([FromHeader(Name = "Authorization")] string authorization)
    {
        var token = authorization.Replace("Bearer ", "");
        return Ok(new { modo = "VULNERABLE", claims = _vulnerable.Validate(token) });
    }

    [HttpPost("seguro/issue")]
    public IActionResult IssueSecure([FromBody] Dictionary<string, string> body)
    {
        var subject = body.GetValueOrDefault("subject", "user1");
        var role = body.GetValueOrDefault("role", "USER");
        return Ok(new
        {
            modo = "SEGURO",
            token = _secure.IssueToken(subject, role)
        });
    }

    [HttpGet("seguro/verify")]
    public IActionResult VerifySecure([FromHeader(Name = "Authorization")] string authorization)
    {
        var token = authorization.Replace("Bearer ", "");
        try
        {
            return Ok(new { modo = "SEGURO", claims = _secure.Validate(token) });
        }
        catch (Exception ex)
        {
            return Unauthorized(new { modo = "SEGURO", valid = false, error = ex.Message });
        }
    }
}

File: sesion5-lab/spring-jwt-attacks/jwt-server/src/main/java/com/example/jwt/web/JwtController.java
package com.example.jwt.web;

import java.util.LinkedHashMap;
import java.util.Map;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.example.jwt.service.SecureJwtService;
import com.example.jwt.service.VulnerableJwtService;

@RestController
@RequestMapping("/api/jwt")
public class JwtController {

    private final VulnerableJwtService vulnerable;
    private final SecureJwtService secure;

    public JwtController(VulnerableJwtService vulnerable, SecureJwtService secure) {
        this.vulnerable = vulnerable;
        this.secure = secure;
    }

    @PostMapping("/vulnerable/issue")
    public Map<String, Object> issueVulnerable(@RequestBody Map<String, String> body) {
        String subject = body.getOrDefault("subject", "user1");
        String role = body.getOrDefault("role", "USER");
        String token = vulnerable.issueToken(subject, role);
        return Map.of(
                "modo", "VULNERABLE (secreto debil en application.yml)",
                "token", token,
                "hint", "Prueba alg=none: eyJhbGciOiJubmUifQ.eyJzdWIiOiJhZG1pbiIsInJvbGUiOiJBRE1JTiJ9.");
    }

    @GetMapping("/vulnerable/verify")
    public ResponseEntity<Map<String, Object>> verifyVulnerable(
            @RequestHeader("Authorization") String authorization) {

        String token = authorization.replace("Bearer ", "");
        var claims = vulnerable.validate(token);
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("modo", "VULNERABLE");
        body.put("claims", claims);
        return ResponseEntity.ok(body);
    }

    @PostMapping("/seguro/issue")
    public Map<String, Object> issueSecure(@RequestBody Map<String, String> body) {
        String subject = body.getOrDefault("subject", "user1");
        String role = body.getOrDefault("role", "USER");
        return Map.of(
                "modo", "SEGURO (HS256, iss/aud validados)",
                "token", secure.issueToken(subject, role));
    }

    @GetMapping("/seguro/verify")
    public ResponseEntity<Map<String, Object>> verifySecure(
            @RequestHeader("Authorization") String authorization) {

        String token = authorization.replace("Bearer ", "");
        try {
            var claims = secure.validate(token);
            return ResponseEntity.ok(Map.of("modo", "SEGURO", "claims", claims));
        } catch (Exception ex) {
            return ResponseEntity.status(401).body(Map.of(
                    "modo", "SEGURO",
                    "valid", false,
                    "error", ex.getMessage()));
        }
    }
}
